File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

n, m, v = map(int,sys.stdin.readline().split())

_dict = {i:[] for i in range(n+1)}
for i in range(m):
    s,e = map(int,sys.stdin.readline().split())
    _dict[s].append(e)
    _dict[e].append(s)

stack, s_dfs, que, q_bfs = [v], [], deque([v]), [] # dfs = stack, s_dfs, bfs = que, q_bfs
visit = [0]*(n+1) # 방문 체크

# dfs
while stack:
    node = stack.pop()
    
    # 방문 체크
    if visit[node]: continue
    visit[node] = 1 

    # stack은 가장 큰 값부터 검사해서, 정점 번호가 작은 것을 먼저 방문하도록 내림차순 정렬
    for i in sorted(_dict[node], reverse=True): 
        if not visit[i]: stack.append(i)
    s_dfs.append(node)

visit = [0]*(n+1) # 방문 초기화
# dfs
while que:
    node = que.popleft()

    # 방문 체크
    if visit[node]: continue
    visit[node] = 1 

    # que는 앞에서부터 pop 시키기 때문에 오름차순 정렬
    for i in sorted(_dict[node]): 
        if not visit[i]: que.append(i)
    q_bfs.append(node)

print(*s_dfs)
print(*q_bfs)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;
import java.util.stream.*;
public class Main {
    static List<Boolean> visit;
    static List<Integer> result;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] NMV = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int N = NMV[0];
        int M = NMV[1];
        int V = NMV[2];


        // 노드 연결
        List<List<Integer>> li = IntStream.rangeClosed(0,N).mapToObj(ArrayList<Integer>::new).collect(Collectors.toList());
        for (int i=0; i<M; i++) {
            int[] node = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            li.get(node[0]).add(node[1]);
            li.get(node[1]).add(node[0]);
        }
        li.forEach(arr-> arr.sort(Comparator.reverseOrder()));

        // dfs
        visit = new ArrayList<>(Collections.nCopies(N+1,false));
        result = new ArrayList<>(N);
        Deque<Integer> stack = new ArrayDeque<>(){{
            add(V);
        }};

        while (!stack.isEmpty()) {
            int now = stack.pollLast();
            if (!visit.get(now)) {
                result.add(now);
                visit.set(now,true);
            }
            for (int num: li.get(now)) {
                if (!visit.get(num)) stack.add(num);
            }
        }
        System.out.println(result.stream().map(String::valueOf).collect(Collectors.joining(" ")));

        // bfs
        li.forEach(Collections::reverse);

        visit = new ArrayList<>(Collections.nCopies(N+1,false));
        result = new ArrayList<>(N);

        Deque<Integer> queue = new ArrayDeque<>(){{
            add(V);
        }};
        visit.set(V,true);
        result.add(V);

        while (!queue.isEmpty()) {
            int now = queue.pollFirst();

            for (int num: li.get(now)) {
                if (!visit.get(num)) {
                    visit.set(num,true);
                    queue.add(num);
                    result.add(num);
                }
            }
        }
        System.out.println(result.stream().map(String::valueOf).collect(Collectors.joining(" ")));
    }
}
